File: nn/rnn/AbstractRNN.py
from abc import ABC, abstractmethod
from nn.Activation import Activation
import numpy as np


class AbstractRNN(ABC):

    def __init__(self, input_size, hidden_size,  output_activation="softmax", hidden_activation='tanh',
                 weight_param=(0, 5), bias_params=(-1, 1), bias_bool=True, training_iterations=3):
        self.Vectors = {}
        self.timestamp = 1
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.output_activation = output_activation
        self.hidden_activation = hidden_activation
        self.weight_params = weight_param
        self.bias_params = bias_params
        self.bias_bool = bias_bool
        self.A = Activation()
        self.Vectors['h0'] = np.zeros(hidden_size)
        self.iterations = training_iterations

    @abstractmethod
    def initiate_weights(self):
        pass

    @abstractmethod
    def feed_forward_one_vect(self, input_vect):
        pass

    @abstractmethod
    def train(self, error):
        pass

    def get_output_abs(self, vect):
        return self.Vectors[vect+str(self.timestamp-1)]

    def feed_forward(self, input_vects):
        assert np.asarray(input_vects).ndim == 2
        self.timestamp = 1
        for vector in input_vects:
            self.feed_forward_one_vect(vector)

File: nn/rnn/AbstractRNN.cpp
#include "AbstractRNN.h"

evo::nn::rnn::AbstractRNN::AbstractRNN(int input_size, int hidden_size, std::string output_activation, std::string hidden_activation,
	std::vector<float> weight_params, std::vector<float> bias_params, bool bias_bool, int training_iterations, float alpha, std::string fp)
	: in_size(input_size), out_size(hidden_size), out_activation(output_activation), hidden_activation(hidden_activation),
	weight_params(weight_params), bias_params(bias_params), bias_bool(bias_bool), alpha(alpha), iterations(training_iterations), fp(fp) {
	vectors["h0"] = evo::empty(hidden_size);
	act = evo::nn::activation(alpha);
}

std::vector<float> evo::nn::rnn::AbstractRNN::feedforward(mtx x) {
	reset_timestamp();
	for (std::vector<float> v : x) {
		feed_forward_one_vect(v);
	}
	return get_output_abs();
}